#2068 added At first download, ask to choose beetween internal and ex…

cdhiraj40 opened the pull request:
Fixes #2068 

**Screenshots** 

https://user-images.githubusercontent.com/75211982/144044600-89777444-c1b1-4119-a607-d72df28ada2c.mp4

gouri-panda commented on core/src/main/res/values/strings.xml:
```diff
@@ -127,8 +127,10 @@
   <string name="tts_pause">pause</string>
   <string name="tts_resume">resume</string>
   <string name="stop">stop</string>
+  <string name="configure_storage_option_title">Download book to internal storage?</string>
```
Is the string value appropriate?

4shutosh replied:
I suggest something like"Choose the storage location"

cdhiraj40 replied:
if we choose "Choose the storage location" then the buttons having strings
will be wrong,
"configure" "yes"
 I followed what @kelson42 said in #2624 
 here->https://github.com/kiwix/kiwix-android/pull/2624#issuecomment-813025562
 I would sure change, but do look once at the comment :)

4shutosh replied:
Oh okay, my bad

gouri-panda replied:
I'm sorry. My bad I forget that comment.

cdhiraj40 replied:
its total fine :) @gouri-panda @4shutosh

gouri-panda commented on app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt:
```diff
@@ -256,17 +258,22 @@ class OnlineLibraryFragment : BaseFragment(), FragmentActivityExtensions {
+          { downloadFile(item.book) },
+          {
+            libraryList.snack(
+              getString(R.string.download_no_space) +
+                "\n" + getString(R.string.space_available) + " " +
```
Maybe concatenate strings would be a better idea?

gouri-panda commented on core/src/main/java/org/kiwix/kiwixmobile/core/utils/SharedPreferenceUtil.kt:
```diff
@@ -172,6 +178,7 @@ class SharedPreferenceUtil @Inject constructor(val context: Context) {
     private const val PREF_NEW_TAB_BACKGROUND = "pref_newtab_background"
     private const val PREF_STORAGE_TITLE = "pref_selected_title"
     private const val PREF_EXTERNAL_LINK_POPUP = "pref_external_link_popup"
+    private const val PREF_SHOW_STORAGE_OPTION = "show_storgae_option"
```
[nit] "pref_storage_option"

cdhiraj40 commented:
@gouri-panda i have done the changes :)

s-ayush2903 commented on app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt:
```diff
@@ -256,17 +258,22 @@ class OnlineLibraryFragment : BaseFragment(), FragmentActivityExtensions {
+          { downloadFile(item.book) },
+          {
+            libraryList.snack(
+              "${getString(R.string.download_no_space)} " +
+                "\n  ${getString(R.string.space_available)}" +
+                it,
```
```suggestion
              "${getString(R.string.download_no_space)}\n${getString(R.string.space_available)} $it",
```
We don't use "+" for joining Strings in Kotlin, this is called _concatenation_, instead interpolation is used to make the code more clean and legible
You can learn about string interpolation from here: https://kotlinlang.org/docs/idioms.html#string-interpolation

cdhiraj40 replied:
oh yes, I do remember doing this but when I committed this I got the error 
``Exceeded max line length`` so I had to take them to the next line which automatically added '+'
any suggestion to resolve this and use your suggested code instead?

s-ayush2903 replied:
Cool, you can fix lint errors via hitting `Ctrl + Alt + L`, and in this case can use kotlin multiline strings:
https://kotlinlang.org/docs/java-to-kotlin-idioms-strings.html#use-multiline-strings

cdhiraj40 replied:
sure will commit these changes soon :)

cdhiraj40 replied:
>Cool, you can fix lint errors via hitting Ctrl + Alt + L, and in this case can use kotlin multiline strings:
https://kotlinlang.org/docs/java-to-kotlin-idioms-strings.html#use-multiline-strings
I tried it but using that doesn't give proper output
![image](https://user-images.githubusercontent.com/75211982/145125345-9c386998-de04-4a1d-81bd-3f4d48551921.png)

cdhiraj40 replied:
>Cool, you can fix lint errors via hitting Ctrl + Alt + L, and in this case can use kotlin multiline strings:
https://kotlinlang.org/docs/java-to-kotlin-idioms-strings.html#use-multiline-strings

@s-ayush2903 I tried it but using that doesn't give proper output
![image](https://user-images.githubusercontent.com/75211982/145125345-9c386998-de04-4a1d-81bd-3f4d48551921.png)

s-ayush2903 replied:
You're supposed to figure that on your own, its not difficult :stuck_out_tongue: , you can try and test on scratches in android studio.
Can refer: https://www.jetbrains.com/help/idea/scratches.html & https://medium.com/@shivam.gosavi340_58315/productivity-hack-androidstudio-kotlin-scratch-file-e17cfde152c6

cdhiraj40 replied:
okay sure :)

s-ayush2903 replied:
Good work. Btw, can you please put a space on the rhs of colon, like it'll be easier to read if written as:
`Space available: xxx` instead `Space available:xxx`

cdhiraj40 replied:
![image](https://user-images.githubusercontent.com/75211982/146151222-3fa4f40f-e94f-43f9-8267-b6047a2fb8ed.png)
done, thanks for suggestion that does makes it easier to read :)

gouri-panda commented:
>@gouri-panda I have done the changes :)

@cdhiraj40  Sorry! My laptop kind of broke right now. but I'll review your PR right after it fixes

cdhiraj40 commented:
please review @gouri-panda

cdhiraj40 commented:
please review @gouri-panda

kelson42 commented:
@gouri-panda I think @cdhiraj40 is expecting your review on this PR since 4 monthes :( Could you please have a look? Would love to bring this in next release.

gouri-panda commented:
@cdhiraj40 @kelson42 sorry! I'll review this PR asap.

gouri-panda reviewed (commented):
@cdhiraj40 We have changed our option to store the zim files a little bit in #2803. Could you update your PR accordingly?

kelson42 commented:
@cdhiraj40 Coukd you please rebase your PR (and resolve conflicts)? Would be great to integrate your fix in 3.6.0.

cdhiraj40 commented:
> @cdhiraj40 Coukd you please rebase your PR (and resolve conflicts)? Would be great to integrate your fix in 3.6.0.

yep sure, done!

cdhiraj40 commented:
> @cdhiraj40 We have changed our option to store the zim files a little bit in #2803. Could you update your PR accordingly?

I am not sure what is changed via #2803 that has relations with this. Could you explain to me what has been changed? Its been a long time since I was active in this project 😅

gouri-panda commented:
@MohitMaliFtechiz Can you help @cdhiraj40 ? Thanks!

MohitMaliFtechiz commented:
> @MohitMaliFtechiz Can you help @cdhiraj40 ? Thanks!

hi @cdhiraj40 , We have added file picker so that for in non play store release version user can select zim file from any storage by its own to read. We have added floating action button in locallibraryfragment.

cdhiraj40 commented:
> > @MohitMaliFtechiz Can you help @cdhiraj40 ? Thanks!
> 
> hi @cdhiraj40 , We have added file picker so that for in non play store release version user can select zim file from any storage by its own to read. We have added floating action button in locallibraryfragment.

Yes I understood that but I was wondering why we need to make any changes here as this PR adds a dialog box asking where to store the zim right?

cdhiraj40 commented:
> > @MohitMaliFtechiz Can you help @cdhiraj40 ? Thanks!
> 
> hi @cdhiraj40 , We have added file picker so that for in non play store release version user can select zim file from any storage by its own to read. We have added floating action button in locallibraryfragment.

Yes, I understood that but I was wondering why we need to make any changes here as this PR adds a dialog box asking where to store the zim right? 
cc @gouri-panda

cdhiraj40 commented:
@gouri-panda I think that has already been done!

cdhiraj40 commented:
@gouri-panda I think this is now good to go? Do let me know!

MohitMaliFtechiz commented on app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt:
```diff
@@ -433,19 +433,22 @@ class OnlineLibraryFragment : BaseFragment(), FragmentActivityExtensions {
-          }
-        )
+        else -> if (sharedPreferenceUtil.showStorageOption) {
+          showStorageConfigureDialog()
```
hi @cdhiraj40 ,

So now we have two version of build specifically for android 11 and above
a) Play-Store version:- App specific without ```MANAGE_EXTERNAL_PERMISSION```
b) F-Droid version:- App specific with ```MANAGE_EXTERNAL_PERMISSION```

So you have place https://github.com/kiwix/kiwix-android/blob/52f7002ba65f080ac2330b0d4b018ca73afe7d13/core/src/main/java/org/kiwix/kiwixmobile/core/utils/SharedPreferenceUtil.kt#L74 and ```Build.VERSION.SDK_INT >= Build.VERSION_CODES.R``` check before showing dialog.

cdhiraj40 replied:
ah got it, okay will do!

kelson42 commented:
@cdhiraj40 Stop merging main branch in your feature branch. Use rebase instead.

cdhiraj40 commented:
> @cdhiraj40 Stop merging main branch in your feature branch. Use rebase instead.

Yes as discussed earlier, I will do that. Its an old commit!

kelson42 commented:
@cdhiraj40 Ok, sorry, though it was recent.

cdhiraj40 commented:
> @cdhiraj40 Ok, sorry, though it was recent.

no worries, I will try it out and see if that works! ignore if you see something wrong here, I will make it right haha!

kelson42 commented:
@MohitMaliFtechiz We need to move quickly forward now. Please feel free superseed this PR so we can at the end merge.

MohitMaliFtechiz reviewed (approved):
Tested and Aproved.
